dali/util/cucontext.cc

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cuda.h>
+#include "dali/error_handling.h"
+#include "dali/util/cucontext.h"
+
+namespace dali {
+
+CUContext::CUContext() : context_{0}, initialized_{false} {
+}
+
+CUContext::CUContext(CUdevice device, unsigned int flags)
+    : device_{device}, context_{0}, initialized_{false} {
+    CUDA_CALL(cuInit(0));
+    CUDA_CALL(cuDevicePrimaryCtxRetain(&context_, device));
+    push();
+    CUdevice dev;
+    CUDA_CALL(cuCtxGetDevice(&dev));
+    initialized_ = true;
+    CUDA_CALL(cuCtxSynchronize());
+}
+
+CUContext::CUContext(CUcontext ctx)
+    : context_{ctx}, initialized_{true} {
+}
+
+CUContext::~CUContext() {
+    if (initialized_) {
+        // cuCtxPopCurrent?
+        CUDA_CALL(cuDevicePrimaryCtxRelease(device_));
+    }
+}
+
+CUContext::CUContext(CUContext&& other)
+    : device_{other.device_}, context_{other.context_},
+      initialized_{other.initialized_} {
+    other.device_ = 0;
+    other.context_ = 0;
+    other.initialized_ = false;
+}
+
+CUContext& CUContext::operator=(CUContext&& other) {
+    if (initialized_) {
+        CUDA_CALL(cuCtxDestroy(context_));
+    }
+    device_ = other.device_;
+    context_ = other.context_;
+    initialized_ = other.initialized_;
+    other.device_ = 0;
+    other.context_ = 0;
+    other.initialized_ = false;
+    return *this;
+}
+
+void CUContext::push() const {
+    CUcontext current;
+    CUDA_CALL(cuCtxGetCurrent(&current));
+    if (current != context_) {
```

Comment:
This check looks like a bad idea - caller has no idea whether the context was actually pushed.
It would be better to have a RAII wrapper around CUcontext which would do a push and a matching pop.

Reply:
I am not sure to get how would this RAII wrapper look like, can you provide me an example?

Reply:
I'm not an expert in CUDA driver API, but I'd try something like:
```
struct CUContextScope {
    CUContextScope(CUcontext ctx) {
        CUcontext current;
        cuCtxGetCurrent(&current)
        success = (current != ctx && cuCtxPushCurrent(ctx) == CUDA_SUCCESS);
    }
    // if we don't mess with context during this object's lifespan, then prev should contain the same context we passed to cuCtxPushCurrent.
    ~CUContextScope() { CUcontext prev; if (success) cuCtxPopCurrent(&prev); }
    bool success;
};
```